### ISSUE TITLE
Object detection with deep learning and OpenCV

### ISSUE THREAD

**Issue by wanghaisheng** [wanghaisheng/awesome-ocr#54]
Deep Learning with OpenCV

http://www.pyimagesearch.com/2017/08/21/deep-learning-with-opencv/
Object detection with deep learning and OpenCV
http://www.pyimagesearch.com/2017/09/11/object-detection-with-deep-learning-and-opencv/

**Comment by wanghaisheng** [comment]
![](http://www.pyimagesearch.com/wp-content/uploads/2017/09/example06_result.jpg)

几周前我们学习了[如何利用 openCV3.3 的 dnn 模块进行图片分类](http://www.pyimagesearch.com/2017/08/21/deep-learning-with-opencv/)。

尽管上一篇文章中介绍了如何给图片打上 ImageNet中的1000种之一的标签，但并不能标出对象在图片中的什么位置。

为了获取图片中对象的bounding box (x, y)坐标，我们需要进行对象识别/对象检测/object detection。


Object detection 不仅能告诉我们图片中是什么东西，也能告诉这个东西在图片中什么位置。

今天的文章中我们会探讨如何利用深度学习技术和openCV来实现 object detection

本文第一部分我们先探讨一下 Single Shot Detectors and MobileNets.


当把这些方法都整合到一起，可以在资源有限的设备上(比如Raspberry  智能手机等等)实现超级快速实时的对象检测。

紧接着我们会探讨如何使用 OpenCV的 dnn 模块来加载一个已经训练好的对象检测模型。

这样我们就可以将输入图片传给模型，然后得到输出的 图片中每个对象的 bouding box x,y 坐标

最后我们来看一下在示例的输入图片上应用 MobileNet Single Shot Detector  的效果。

随后的一篇文章中我们会改进现有的代码实现实时视频流的处理。

**Comment by wanghaisheng** [comment]
## Single Shot Detectors for object detection
![](http://www.pyimagesearch.com/wp-content/uploads/2017/09/single_shot_detector_example.jpg)

当大家讨论基于深度学习的对象检测技术，那么主要有如下三种技术：

*    [Faster R-CNNs (Girshick et al., 2015)](https://arxiv.org/abs/1506.01497)
*    [You Only Look Once (YOLO) (Redmon and Farhadi, 2015)]()
*    [Single Shot Detectors (SSDs) (Liu et al., 2015)]()

Faster R-CNN 是最为人所知的基于深度学习的对象检测技术；但是，该技术难以理解(特别是对于深度学习的初学者来讲)，难于实现，训练起来很麻烦。

另外，即使是 R-CNN 中最快的实现 该算法其实也挺慢的，(where the “R” stands for “Region Proposal”) the algorithm can be quite slow, on the order of 7 FPS.

如果只是单纯追求速度，那么我们倾向于选择YOLO这种速度更快，大概在一块Titan X GPU上能达到 40-90 FPS 的算法。最快的 YOLO 变种甚至可以达到 155 FPS。

YOLO 算法的问题在于 准确性有待提高。

SSD,起初是 Google 提出来的，是Faster R-CNN 和YOLO 算法两者之间的折中。这种算法比 R-CNN 更加直截了当。

如果选择合适的这种网络的变种，甚至我们可以得到比[Girshick et al. at 22-46](https://arxiv.org/abs/1506.01497) 更高的 FPS 吞吐量.在准确性上  SSD 也要优于 YOLO。更多详细信息请阅读, [Liu et al](https://arxiv.org/abs/1512.02325).

**Comment by wanghaisheng** [comment]
## MobileNets: Efficient (deep) neural networks
![](http://www.pyimagesearch.com/wp-content/uploads/2017/09/mobilenet_depthwise.png)

在训练对象检测模型/网络的时候 通常我们会选择使用一种现成的网络架构，比如 VGG ResNet，将其用在对象检测pipeline的内部。问题在于这些网络的大小都会在200-500MB 左右。

受限于sheer size and resulting number of computation，诸如此类的网络架构在资源有限的设备上是不适用的。

取而代之的，我们使用的是 [MobileNets (Howard et al., 2017)](https://arxiv.org/abs/1704.04861),这是 Google 研发人员发表的另外一篇论文。我们将其称之为“MobileNets”  是因为他们是专门为智能手机这样的资源有限的设备而设计的。MobileNets 与传统的 CNN算法的区别之处在于使用了 depthwise separable convolution (Figure 2 above).

depthwise separable convolution 的基本原理可以将 convolution 分为两阶段：

1.    A 3×3 depthwise convolution.
 2.   Followed by a 1×1 pointwise convolution.

这样我们就可以减少网络中参数的数量。


引入的问题在于我们牺牲了准确性- “MobileNets” 不如那些较大的网络准确性高。。。
但它的性价比更高。

要了解有关 MobileNets ，请阅读 [Howard et al](https://arxiv.org/abs/1704.04861).

**Comment by wanghaisheng** [comment]
## Combining MobileNets and Single Shot Detectors for fast, efficient deep-learning based object detection

如果我们将 MobileNet 和 SSD 整合在一起，我们就能得到一种速度既快又高效的基于深度学习的对象检测技术。

这篇文章中使用的模型是 [Howard et al](https://github.com/Zehaos/MobileNet)提出的 Tensorflow模型的 Caffe版本，是由chuanqi305训练好的 (see [GitHub](https://github.com/chuanqi305/MobileNet-SSD)).

首先在 COCO 数据集上训练得到 MobileNet SSD ，然后利用 PASCAL VOC 调优 达到 72.7% mAP 

这样我们就可以在图片中检测20种对象(背景要算一种)，包括 airplanes, bicycles, birds, boats, bottles, buses, cars, cats, chairs, cows, dining tables, dogs, horses, motorbikes, people, potted plants, sheep, sofas, trains, and tv monitors.

**Comment by wanghaisheng** [comment]
## Deep learning-based object detection with OpenCV

这部分介绍如何利用 opencv的 dnn模块 +MobileNet SSD来构建我们自己的对象检测器。

你可以下载源代码、训练好的模型、示例图片，在自己本地进行测试。

Let’s go ahead and get started building our deep learning object detector using OpenCV.

Open up a new file, name it deep_learning_object_detection.py , and insert the following code:

**Comment by wanghaisheng** [comment]
Traditional object detection is accomplished using a sliding window an image pyramid, like in Histogram of Oriented Gradients. Deep learning-based object detectors do end-to-end object detection. The actual inner workings of how SSD/Faster R-CNN work are outside the context of this post, but the gist is that you can divide an image into a grid, classify each grid, and then adjust the anchors of the grid to better fit the object. This is a huge simplification but it should help point you in the right direction.
